src/d3d9/d3d9_stateblock.h

```diff
@@ -211,11 +216,19 @@ namespace dxvk {
       if (m_captures.flags.test(D3D9CapturedStateFlag::VertexBuffers)) {
         for (uint32_t idx : bit::BitMask(m_captures.vertexBuffers.dword(0))) {
           const auto& vbo = src->vertexBuffers[idx];
-          dst->SetStreamSource(
-            idx,
-            vbo.vertexBuffer.ptr(),
-            vbo.offset,
-            vbo.stride);
+          if constexpr (!IgnoreStreamOffset) {
+            dst->SetStreamSource(
+              idx,
+              vbo.vertexBuffer.ptr(),
+              vbo.offset,
+              vbo.stride);
+          } else {
+            // For whatever reason, D3D9 doesn't capture the stream offset
+            dst->SetStreamSourceWithoutOffset(
```

Comment:
Where does the offset we apply to the context come from if it is never captured?

Reply:
Probably whatever offset was set on the device at the time.

Reply:
Okay it's pretty weird:

It **only** ignores the offset when calling `IDirect3DStateBlock9::Capture`.
That means, that both the initial capture when creating a state block using `IDirect3DDevice9::CreateStateBlock` as well as capturing the stream source by calling `IDirect3DDevice9::SetStreamSource` in between `IDirect3DDevice9::BeginStateBlock` and `IDirect3DDevice9::EndStateBlock` will capture the offset as one would normally expect.

`IDirect3DStateBlock9::Apply` only touches the stream sources if the state block has been created using `IDirect3DDevice9::CreateStateBlock(D3DSBT_VERTEXSTATE)` or by calling `IDirect3DDevice9::SetStreamSource` in between `IDirect3DDevice9::BeginStateBlock` and `IDirect3DDevice9::EndStateBlock`. Both of those capture the offset just fine.

So the only thing that does not work is **updating** the offset of a stream source that is capturing in a state block by calling `Capture`. So to answer your question: it's not strictly correct that the offset is "never captured" and the offset we apply comes from capturing it earlier.

I hope that was somewhat understandable. 😅